src/riscv.c

```diff
@@ -211,6 +220,74 @@ static void *t2c_runloop(void *arg)
 }
 #endif
 
+#if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
+static void map_file(char **ram_loc, const char *name)
+{
+    int fd = open(name, O_RDONLY);
+    if (fd < 0) {
+        fprintf(stderr, "could not open %s\n", name);
+        exit(2);
+    }
+
+    /* get file size */
+    struct stat st;
+    fstat(fd, &st);
+
+#if HAVE_MMAP
+    /* remap to a memory region */
+    *ram_loc = mmap(*ram_loc, st.st_size, PROT_READ | PROT_WRITE,
+                    MAP_FIXED | MAP_PRIVATE, fd, 0);
+    if (*ram_loc == MAP_FAILED) {
+        perror("mmap");
+        close(fd);
+        exit(2);
+    }
+#else
+    /* calloc and load data to a memory region */
+    *ram_loc = calloc(st.st_size, sizeof(uint8_t));
+    if (!*ram_loc) {
+        perror("calloc");
+        close(fd);
+        exit(2);
+    }
+    if (read(fd, *ram_loc, st.st_size) != st.st_size) {
+        perror("read");
+        close(fd);
+        exit(2);
+    }
```

Comment:
When `read()` fails, we should free `*ram_loc` here.
Additionally, the
```
close(fd);
exit(2);
```
is repeated multiple times here. Maybe we can move them to the end and use `goto` for error handling?